pyteal/ast/abirouter.py

```diff
@@ -0,0 +1,213 @@
+            else Txn.application_args.length() == Int(0)
+        )
+        approvalConds.append(methodOrBareCondition)
+
+        # Check the existence of OC.CloseOut
+        closeOutExist = any(map(lambda x: x == OnComplete.CloseOut, onCompletes))
+        # Check the existence of OC.ClearState (needed later)
+        clearStateExist = any(map(lambda x: x == OnComplete.ClearState, onCompletes))
+        # Ill formed report if app create with existence of OC.CloseOut or OC.ClearState
+        if creation and (closeOutExist or clearStateExist):
+            raise TealInputError(
+                "OnComplete ClearState/CloseOut may be ill-formed with app creation"
+            )
+        # if OC.ClearState exists, add method-or-bare-condition since it is only needed in ClearStateProgram
+        if clearStateExist:
+            clearStateConds.append(methodOrBareCondition)
+
+        # Check onComplete conditions for approvalConds, filter out *ClearState*
+        approvalOcConds: List[Expr] = [
+            Txn.on_completion() == oc
+            for oc in onCompletes
+            if oc != OnComplete.ClearState
+        ]
+
+        # if approval OC condition is not empty, append Or to approvalConds
+        if len(approvalOcConds) > 0:
+            approvalConds.append(Or(*approvalOcConds))
+
+        # what we have here is:
+        # list of conds for approval program on one branch: creation?, method/bare, Or[OCs]
+        # list of conds for clearState program on one branch: method/bare
+        return approvalConds, clearStateConds
+
+    @staticmethod
+    def wrapHandler(isMethod: bool, branch: Union[SubroutineFnWrapper, Expr]) -> Expr:
+        exprList: List[Expr] = []
+        if not isMethod:
+            if (
+                isinstance(branch, Seq)
+                and not branch.has_return()
+                and branch.type_of() == TealType.none
+            ):
+                exprList.append(branch)
+            elif (
+                isinstance(branch, SubroutineFnWrapper)
+                and branch.has_return()
+                and branch.type_of() == TealType.none
+            ):
+                exprList.append(branch())
+            else:
+                raise TealInputError(
+                    "For bare app call: should only register Seq (with no ret) or Subroutine (with ret but none type)"
+                )
+        else:
+            if isinstance(branch, SubroutineFnWrapper) and branch.has_return():
+                # TODO need to encode/decode things
+
+                execBranch: Expr = branch(
+                    *[
+                        Txn.application_args[i + 1]
+                        for i in range(len(branch.subroutine.implementationParams))
+                    ]
+                )
+
+                exprList.append(
+                    Log(Concat(RETURN_EVENT_SELECTOR, execBranch))
+                    if branch.type_of() != TealType.none
+                    else execBranch
+                )
```

Comment:
This part is kinda waiting on ABI, if we can implement encoding/decoding for arguments and types without blocked by ABI, that would be nice.